Add sox handler. fixes #106

enclarify opened the pull request:
**Issue this PR fixes (if applicable):** #106

**Description:**
Adds a SoX playback handler. Additionally it refactors the basehandler and vlchandler to reuse code common to all handlers

**Checklist:**
  - [x] Update CHANGELOG.md
  - [x] Update Wiki (new platform, installation, configuration changes etc)
  - [x] PR Location (usually `master` for hotfixes, `dev` for general development)
  - [x] rebase on `dev` and change the prints to logging
  - handling (m3u8) playlists - https://pypi.python.org/pypi/m3u8/ - we might actually do this on AlexaPi level and pass only real streams to the playback handlers ... we'll see ... doesn't need to be a part of this PR

dgonano commented:
how come this doesn't have the PR template?

enclarify commented:
Because I only realized after submission that it was expected that the entire template remain in the description. Will amend

dgonano commented:
No worries. It just has a checklist in to make sure everything gets done for the merge. e.g. updating the changelog and wiki ;) 

Also as a side node, I saw an email with a merge commit in it but I don't see it here so I assume it was deleted?
Anyway please do a rebase instead of a merge when you can to help keep the history clean and PR easy to follow.

enclarify commented:
Yeah, I prefer rebase as well. I edited a merge conflict via the github web UI when I ran out to dinner and it did the merge commit. Will force push that away in the morning

dgonano commented:
I only did a merge on your triggering @renekliment because I didn't want to rebase your repo and force push :p 

I guess it only effects the dev and master if you do a fast forward merge or merge commit. If you do a squash and merge (like we seem to be) it all gets condensed to a single commit anyway so I guess its not an issue.

renekliment commented on src/config.template.yaml:
```diff
@@ -26,6 +26,10 @@ sound:
 
+  # Some devices clip off the beginning/end of audio files being played back. The
+  # value is the number of seconds of silence to pad the beginning and end of the audio
+  playback_padding: 0
```
Good idea! Although the cause should be fixed. Now it feels like linux/systemd when we're supposed to fix someone else's mess. Well, since there's probably no easy way to fix the damn CHIP audio, I'm glad we have this!

renekliment replied:
You need to mention this in the wiki/Configuration changes.

renekliment commented on src/alexapi/playback_handlers/vlchandler.py:
```diff
@@ -191,16 +115,13 @@ def state_callback(self, event, player): # pylint: disable=unused-argument
 
-			rThread = threading.Thread(target=self.__callback_report, args=report[state])
-			rThread.start()
+			report[state]()
```
Doesn't synchronous reporting have noticeable slowdown?

enclarify replied:
If you look at commit 4e42544ba121c1a367d7343a4386b9d6b42144af you will see that the reporting to AVS is still async. Only the platform calls are blocking now which is likely the correct behavior. The comment on that commit says a little more

renekliment replied:
Ah, I see, missed that. Yes, this is better :+1:

renekliment commented on src/alexapi/playback_handlers/soxhandler.py:
```diff
@@ -0,0 +1,101 @@
+            self.report_finish()
+
+    def on_stop(self):
+        kill = subprocess.Popen(['pkill', '-KILL', 'sox'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
This looks like it kills all `sox` instances, not just ours. We can't do it like this.

enclarify replied:
yeah, will cleanup and kill just ours

renekliment commented on src/alexapi/playback_handlers/basehandler.py:
```diff
@@ -1,41 +1,171 @@
 		pass
 
-	@abstractmethod
+	def report_play(self, debug_msg=None):
```
I don't really like the zillion `report_` methods, but meh, I guess it's still fine.
Also, with the logging stuff, it will be shorter.

enclarify replied:
The report_ methods are superior to depending on each subclass to report the states and activity correctly. Encapsulation behind the report_ methods makes playback handler implementations simpler. This prevents errors like those identified in #117

renekliment replied:
Alright.

renekliment commented:
I've edited the description to include the template.

Also, I hope that when you say _provisional_, you mean _preliminary_ and will not close & open a new / replace this PR :-)

Looking forward to seeing the comments addressed & testing this.

renekliment commented on src/alexapi/playback_handlers/soxhandler.py:
```diff
@@ -0,0 +1,101 @@
+        # SoX can't play file URLs
+        audio_file = item.url.replace('file://', '')
+
+        sox_cmd = ['sox', '-q', audio_file]
```
You use the command `sox` instead of `play`. How come this works? (again, haven't tested)

enclarify replied:
The `play` and `rec` commands are just shortcuts that delegate back to the main sox code. All of the same functionality is available in the main `sox` executable. Furthermore the output devices that can be changed with the `-t` switch of the `sox` command have to be set with environment variables when using `play` and `rec`. It's simpler to use the sox command directly plus it keeps the code from having to use the `shell=True` attribute of Popen.

renekliment replied:
I see, all clear now.

enclarify commented:
In regards to the comment about the `on_` methods. The point is not just for debug logging nor just for design pattern style. The code can definitely be written differently but the methods implemented by each playback handler need to be wrapped, not extended, in many cases to effectively handle the concurrency in the codebase. For example look at the stop() method. In the current implementation the playback lock is not released. IMHO playback handler implementations should be ignorant of the concurrency details of the calling code

renekliment commented:
@enclarify Point taken.

enclarify commented:
In regards to pysox, not sure there is a huge advantage to using it and in some cases it can't be used. For example once the `pkill` that that's used in the stop method is fixed the pid from the Popen will be needed and pysox doesn't pass that back. On top of that it's internally inconsistent in how it works where sometimes it concats args and Popens with a shell and other times it does not. For the purposes of AlexaPi it seems simpler to work with `sox` and `soxi` directly

enclarify commented on src/scripts/inc/archlinux.sh:
```diff
@@ -1,8 +1,19 @@
     pacman -Sy
-    pacman -S git python2 python2-pip swig alsa-lib alsa-utils libpulse vlc memcached gcc --noconfirm --needed
+    pacman -S git python2 python2-pip swig alsa-lib alsa-utils libpulse vlc libid3tag memcached gcc make --noconfirm --needed
+    # the sox package for arch linux does not have does not have MP3 support
```
You are correct, adding libmad solves this

renekliment commented:
Sox supposedly supports m3u:
```
renekl@piggy ~ % sox -h | grep m3u              
PLAYLIST FORMATS: m3u pls
```
and there's some stuff about it in documentation. And you can do 
```
sox -t mp3 http://www.ndr.de/resources/metadaten/audio/m3u/ndrloop5.m3u -t alsa`
```
so it looks inside, but I'm not sure if it really handles `m3u8` or is capable of grabbing a simple URL only. Also, it doesn't work without the `-t mp3`, but the playlist can contain pretty much anything.

This is here just for discussion - as I said, I'd be happy to merge it without messing with this and do this later.

enclarify commented:
SoX does handle m3u well. However from my manual testing it doesn't work with m3u8. I tried various things to try to get it to think it was just a m3u or mp3 without success. From what I have read m3u8 is just supposed to be UTF-8 encoded m3u but it also sounds like there can be extra embedded metadata. Asking Alexa to "play some jazz music" returns a m3u8 stream served out from cloudfront. It doesn't seem like the VLC support for m3u8 is that great either because I was able to occasional hang the AlexaPi process during testing even though the stream did start to play. It would be cool to find a SoX way of handling this

renekliment commented:
I see. Okay, this seems like it's gonna require a lot of testing and doing stuff, so let's do it later.

renekliment commented on src/alexapi/playback_handlers/soxhandler.py:
```diff
@@ -0,0 +1,107 @@
+		if item.offset:
+			sox_cmd.extend(['trim', self.__calculate_offset(item.offset)])
+
+		sox_cmd.extend(['pad', self.playback_padding, self.playback_padding])
```
Why is `self.playback_padding` here twice?

enclarify replied:
Some SoX effects like padding can be applied both at the beginning and end of the audio. In this case it seemed in testing that padding the beginning and end of audio playback offered a cleaner transition than not. Given this padding effect is generally a hack for platforms not performing as well as others yet it seemed reasonable

renekliment replied:
Okay. Thanks for the explanation.

renekliment commented on CHANGELOG.md:
```diff
@@ -12,14 +12,15 @@ Please update your config according to the [Configuration changes] section on th
     - Each trigger can be enabled / disabled. Voice triggering is therefore now optional (although enabled by default).
     - Pocketsphinx's _threshold_ is now configurable.
 - Use Python logging instead of prints to stdout
+- Refactored playback handlers. Common functionality moved into BaseHandler
```
Remove this please. `CHANGELOG.md` is for users.

renekliment commented on CHANGELOG.md:
```diff
@@ -12,14 +12,15 @@ Please update your config according to the [Configuration changes] section on th
 	- logging level
 - New _hyperion_ platform. Allows status visualization with [Hyperion](https://hyperion-project.org).
-
+- SoX playback handler now configurable in addition to VLC
```
```markdown
- SoX playback handler in addition to the existing VLC handler 
    - This should fix/improve issues with audio on Orange Pi a CHIP (see also the `playback_padding` config option)
    - TuneIn support is experimental and will be improved in the future
```

renekliment commented:
Thanks for the change log change.

Please:
- [ ] in the config template mention that `playback_padding` is for SoX handler only
- [ ] rebase on `dev`

and I'll merge this.

Thanks!

enclarify commented:
Good suggestion about adding a note that the `playback_padding` only works with the sox handler. I added that note back in commit 3055d2a. Also rebased on dev

renekliment commented:
@enclarify Thank you for the changes!

I've made one last test and I've got one more thing I've missed earlier :disappointed: 
- [ ] The `report_` methods in `basehandler.py` should log debug messages regardless of the `debug_msg` parameter. For example now I don't get the message that playback stopped when using the SoX handler. I think this is general for all the methods / handlers. The methods are often called without the parameter, but I'd like to see at least the general messages.
- If you're doing changes, it would be neat if you could squash 0d58ac5ce339199a13a6df8801c9ea2e25ae01e4 into the commit that adds the SoX handler.

I truly believe this would be the last set of changes. Thank you.

enclarify commented:
Latest comments have been addressed

renekliment commented:
Dope! Merging! Thank you!
